runtime/lua/vim/treesitter/highlighter.lua

```diff
@@ -328,45 +335,89 @@ end
+local function add_mark(
+  m,
+  buf,
+  range_start_row,
+  range_start_col,
+  range_end_row,
+  range_end_col,
+  next_marks
+)
   local cur_start_l = m.start_line
   local cur_start_c = m.start_col
-  if cur_start_l < line then
-    cur_start_l = line
-    cur_start_c = 0
+  if cmp_lt(cur_start_l, cur_start_c, range_start_row, range_start_col) then
+    cur_start_l = range_start_row
+    cur_start_c = range_start_col
   end
 
   local cur_opts = m.opts
-  if cur_opts.end_line >= line + 1 then
+  if cmp_lt(range_end_row, range_end_col, cur_opts.end_line, cur_opts.end_col) then
     cur_opts = vim.deepcopy(cur_opts, true)
-    cur_opts.end_line = line + 1
-    cur_opts.end_col = 0
+    cur_opts.end_line = range_end_row
+    cur_opts.end_col = range_end_col
     table.insert(next_marks, m)
   end
 
-  local empty = cur_opts.end_line < cur_start_l
-    or (cur_opts.end_line == cur_start_l and cur_opts.end_col <= cur_start_c)
-  if cur_start_l <= line and not empty then
+  if cmp_lt(cur_start_l, cur_start_c, cur_opts.end_line, cur_opts.end_col) then
     api.nvim_buf_set_extmark(buf, ns, cur_start_l, cur_start_c, cur_opts)
   end
 end
 
 ---@param self vim.treesitter.highlighter
 ---@param win integer
 ---@param buf integer
----@param line integer
+---@param range_start_row integer
+---@param range_start_col integer
+---@param range_end_row integer
+---@param range_end_col integer
 ---@param on_spell boolean
 ---@param on_conceal boolean
-local function on_line_impl(self, win, buf, line, on_spell, on_conceal)
-  self._conceal_checked[line] = self._conceal_line and true or nil
+local function on_range_impl(
+  self,
+  win,
+  buf,
+  range_start_row,
+  range_start_col,
+  range_end_row,
+  range_end_col,
+  on_spell,
+  on_conceal
+)
+  if self._conceal_line then
+    range_start_col = 0
+    if range_end_col ~= 0 then
+      range_end_row = range_end_row + 1
+      range_end_col = 0
+    end
+  end
+  for i = range_start_row, range_end_row - 1 do
+    self._conceal_checked[i] = self._conceal_line or nil
+  end
   self:for_each_highlight_state(win, function(state)
     local root_node = state.tstree:root()
-    local root_start_row, _, root_end_row, _ = root_node:range()
+    local root_start_row, root_start_col, root_end_row, root_end_col = root_node:range()
+
+    --- @type boolean
+    local root_intersects
+
+    local range_empty = cmp_eq(range_start_row, range_start_col, range_end_row, range_end_col)
+    local root_empty = cmp_eq(root_start_row, root_start_col, root_end_row, root_end_col)
+    if not range_empty and not root_empty then
+      root_intersects = cmp_lt(range_start_row, range_start_col, root_end_row, root_end_col)
+        and cmp_lt(root_start_row, root_start_col, range_end_row, range_end_col)
+    else
+      root_intersects = cmp_lte(range_start_row, range_start_col, root_end_row, root_end_col)
+        and cmp_lte(root_start_row, root_start_col, range_end_row, range_end_col)
+    end
```

Comment:
It's a bit unfortunate if this needs to be this much more complex. If the range becomes empty (as a result of the fold clipping) can't we skip the `for_each` completely? 

Also when a particular tree root is empty, we can just skip the tree at that point, right?